README.md
# TSLint Config Airbnb

[![NPM version](https://img.shields.io/npm/v/tslint-config-airbnb.svg?style=flat)](https://www.npmjs.com/package/tslint-config-airbnb)
[![Downloads](http://img.shields.io/npm/dm/tslint-config-airbnb.svg?style=flat)](https://npmjs.org/package/tslint-config-airbnb)

> A [TSLint config](https://palantir.github.io/tslint/usage/configuration/) for [Airbnb JavaScript Style Guide](https://github.com/airbnb/javascript)

---

## This package is deprecated

### Migrate to eslint

1. Remove `tslint.json`

2. Remove lines from `tsconfig.json` if included

```json
    "plugins": [
      {
        "name": "typescript-tslint-plugin"
      }
    ],
```

3. Replace packages

```bash
$ npm uninstall tslint tslint-config-airbnb
$ npm uninstall typescript-tslint-plugin
$ npm install --save-dev eslint @typescript-eslint/parser @typescript-eslint/eslint-plugin eslint-config-airbnb
$ npx install-peerdeps --dev eslint-config-airbnb
```

4. Create `.eslintrc.json`

```json
{
  "extends": ["airbnb"],
  "plugins": ["@typescript-eslint"],
  "parser": "@typescript-eslint/parser",
  "rules": {
    "import/no-unresolved": 0,
    "react/jsx-filename-extension": {
      "extensions": [
        ".jsx",
        ".tsx"
      ]
    }
  }
}
```

With webpack

```json
{
  :
  "rules": {
    :
    "import/extensions": [".js", ".json", ".jsx", ".ts", ".tsx"],
    :
  },
  "settings": {
    "import/resolver": {
      "webpack": {
        "config": "webpack.config.js"
      }
    }
  }
  :
}
```

5. if you are using Visual Studio Code, add to `vscode's settings.json`

```json
{
  "[typescript]": {
    "editor.formatOnSave": false,
  },
  "eslint.autoFixOnSave": true,
  "eslint.validate": [
    "javascript",
    "javascriptreact",
    { "language": "typescript", "autoFix": true },
    { "language": "typescriptreact", "autoFix": true }
  ]
}
```

---

## Installation

```sh
npm install tslint-config-airbnb --save-dev
```

## Usage

In `tslint.json`:

```json
{
  "extends": "tslint-config-airbnb"
}
```

### Rules

- [tslint](https://www.npmjs.com/package/tslint)
- [tslint-consistent-codestyle](https://www.npmjs.com/package/tslint-consistent-codestyle)
- [tslint-eslint-rules](https://www.npmjs.com/package/tslint-eslint-rules)
- [tslint-microsoft-contrib](https://www.npmjs.com/package/tslint-microsoft-contrib)

## Versioning

```
+----- Major version is synchronize with tslint's major version.
| +--- Minor version has BREAKING CHANGE and feat.
| | +- Patch version has patch.
| | |
x.x.x
```

## License

Apache 2.0


CHANGELOG.md
<a name="5.11.2"></a>
## [5.11.2](https://github.com/progre/tslint-config-airbnb/compare/5.11.1...5.11.2) (2019-09-19)



<a name="5.11.1"></a>
## [5.11.1](https://github.com/progre/tslint-config-airbnb/compare/5.11.0...5.11.1) (2018-11-15)


### Bug Fixes

* wrong rule configuration ([3abd09c](https://github.com/progre/tslint-config-airbnb/commit/3abd09c)), closes [#56](https://github.com/progre/tslint-config-airbnb/issues/56)



<a name="5.11.0"></a>
# [5.11.0](https://github.com/progre/tslint-config-airbnb/compare/5.10.0...5.11.0) (2018-09-01)


### Bug Fixes

* remove deprecated rule "no-unused-variable" ([5b54968](https://github.com/progre/tslint-config-airbnb/commit/5b54968)), closes [#54](https://github.com/progre/tslint-config-airbnb/issues/54)

and a PR

* https://github.com/progre/tslint-config-airbnb/pull/50


### Features

* https://github.com/progre/tslint-config-airbnb/pull/49


<a name="5.10.0"></a>
# [5.10.0](https://github.com/progre/tslint-config-airbnb/compare/5.9.2...5.10.0) (2018-08-18)


### Bug Fixes

* allow-destructuring on no-this-assignment ([d80f2f0](https://github.com/progre/tslint-config-airbnb/commit/d80f2f0)), closes [#47](https://github.com/progre/tslint-config-airbnb/issues/47)


### Features

* update dependencies ([0cd45bf](https://github.com/progre/tslint-config-airbnb/commit/0cd45bf)), closes [#52](https://github.com/progre/tslint-config-airbnb/issues/52) [#53](https://github.com/progre/tslint-config-airbnb/issues/53)

and some PRs

* https://github.com/progre/tslint-config-airbnb/pull/40
* https://github.com/progre/tslint-config-airbnb/pull/41
* https://github.com/progre/tslint-config-airbnb/pull/42
* https://github.com/progre/tslint-config-airbnb/pull/48

<a name="5.9.0"></a>
# [5.9.0](https://github.com/progre/tslint-config-airbnb/compare/5.8.0...5.9.0) (2018-05-19)


### Features

* add whitespace rules ([1c9864d](https://github.com/progre/tslint-config-airbnb/commit/1c9864d))



<a name="5.8.0"></a>
# [5.8.0](https://github.com/progre/tslint-config-airbnb/compare/5.7.0...5.8.0) (2018-03-19)


### Features

* remove strict-boolean-expressions [#29](https://github.com/progre/tslint-config-airbnb/issues/29) ([50d7342](https://github.com/progre/tslint-config-airbnb/commit/50d7342))



<a name="5.7.0"></a>
# [5.7.0](https://github.com/progre/tslint-config-airbnb/compare/5.6.0...5.7.0) (2018-02-21)


### Features

* add curly for 16.1 ([5143689](https://github.com/progre/tslint-config-airbnb/commit/5143689))



<a name="5.6.0"></a>
# [5.6.0](https://github.com/progre/tslint-config-airbnb/compare/5.5.0...5.6.0) (2018-02-07)


### Bug Fixes

* resolve interfere with 15.7 ([7396b2b](https://github.com/progre/tslint-config-airbnb/commit/7396b2b))



<a name="5.5.0"></a>
# [5.5.0](https://github.com/progre/tslint-config-airbnb/compare/5.4.2...5.5.0) (2018-01-24)


### Features

* replace no-var-self with no-this-assignment ([397e531](https://github.com/progre/tslint-config-airbnb/commit/397e531))
* add peerDependencies ([7b620a5](https://github.com/progre/tslint-config-airbnb/commit/7b620a5))



<a name="5.4.2"></a>
## [5.4.2](https://github.com/progre/tslint-config-airbnb/compare/5.4.1...5.4.2) (2017-11-15)


### Bug Fixes

* allow $method ([a980f53](https://github.com/progre/tslint-config-airbnb/commit/a980f53)), closes [#24](https://github.com/progre/tslint-config-airbnb/issues/24)



<a name="5.4.1"></a>
# [5.4.1](https://github.com/progre/tslint-config-airbnb/compare/5.3.0...5.4.1) (2017-11-15)


### Features

* add no-else-after-return for 16.3 ([36c678a](https://github.com/progre/tslint-config-airbnb/commit/36c678a))



<a name="5.3.1"></a>
## [5.3.1](https://github.com/progre/tslint-config-airbnb/compare/5.3.0...5.3.1) (2017-11-13)


### Bug Fixes

* update dependencies ([c91dcff](https://github.com/progre/tslint-config-airbnb/commit/c91dcff))





sample/index.ts
class SampleClass {
  member: {};

  method ( ) {
    this.member = {};
    const shorthand = 1;
    const isValid: boolean = null;

    // bad
    if (isValid === true) {
      // ...
    }

    // good
    if (isValid) {
      // ...
    }

    new Promise((resolve, reject) => {
    });

    return {
      a: 1,
      shorthand,
      b: 1
    };
  }
}

function main() {
  const sample = new SampleClass();
  sample.method();
}

main();

function rule15_3(isValid: boolean, name: string, collection: Object[]) {
  // bad
  if (isValid === true) {
    // ...
  }

  // good
  if (isValid) {
    // ...
  }

  // bad
  if (name) {
    // ...
  }

  // good
  if (name !== '') {
    // ...
  }

  // bad
  if (collection.length) {
    // ...
  }

  // good
  if (collection.length > 0) {
    // ...
  }
}

function rule15_7_bad(a: Object, b: Object, c: Object) {
  const foo = a ? a : b;
  const bar = c ? true : false;
  const baz = c ? false : true;
}

function rule15_7_good(a: Object, b: Object, c: Object) {
  const foo = a || b;
  const bar = !!c;
  const baz = !c;
}

function rule16_1(test: boolean) {
  // bad
  if (test)
    return false;

  // good
  if (test) return false;

  // good
  if (test) {
    return false;
  }

  // bad
  function foo() { return false; }

  // good
  function bar() {
    return false;
  }
}


tslint.js
const path = require('path');

module.exports = {
  rulesDirectory: [
    path.join(path.dirname(require.resolve('tslint-consistent-codestyle')), './'),
    path.join(path.dirname(require.resolve('tslint-eslint-rules')), 'dist/rules'),
    path.join(path.dirname(require.resolve('tslint-microsoft-contrib')), './'),
  ],
  rules: { // based on https://github.com/airbnb/javascript/tree/74fb34f08d4f1961169e9c37c9036a5b42aa7bf1
    'prefer-const': true, // 2.1, 13.1
    'no-var-keyword': true, // 2.2
    'object-literal-shorthand': true, // 3.3, 3.4
    'object-shorthand-properties-first': true, // 3.5
    'object-literal-key-quotes': [true, 'as-needed'], // 3.6
    'prefer-array-literal': true, // 4.1
    quotemark: [
      true,
      'single',
      'avoid-escape',
      'avoid-template',
      'jsx-double',
    ], // 6.1, 6.5
    'prefer-template': true, // 6.3
    'no-eval': true, // 6.4
    'function-constructor': true, // 7.10
    'space-before-function-paren': [
      true,
      {
        anonymous: 'always', // 7.11
        named: 'never', // 19.3
      },
    ],
    'no-parameter-reassignment': true, // 7.12
    align: [
      true,
      'arguments',
      'parameters',
    ], // 7.15
    'ter-prefer-arrow-callback': [true], // 8.1
    'arrow-parens': false, // 8.2
    'ter-arrow-parens': [
      true,
      'as-needed',
      { 'requireForBlockBody': true },
    ], // 8.4
    'no-duplicate-imports': true, // 10.4
    'one-variable-per-declaration': [true, 'ignore-for-loop'], // 13.2
    'increment-decrement': true, // 13.6
    'triple-equals': [true, 'allow-null-check'], // 15.1
    'no-boolean-literal-compare': true, // 15.3
    curly: [true, 'ignore-same-line'], // 16.1
    'brace-style': [
      true,
      '1tbs',
      { allowSingleLine: true },
    ], // 16.2
    'no-else-after-return': true, // 16.3
    'comment-format': [true, 'check-space'], // 18.3
    indent: [true, 'spaces'], // 19.1
    'ter-indent': [
      true,
      2,
      { 'SwitchCase': 1 },
    ], // 19.1
    whitespace: [
      true,
      'check-branch', // 19.3
      'check-decl', // 19.4
      'check-operator', // 19.4
      'check-preblock', // 19.2
      'check-separator', // 19.14
    ],
    eofline: true, // 19.5
    'space-in-parens': [true, 'never'], // 19.9
    'array-bracket-spacing': [true, 'never'], // 19.10
    'object-curly-spacing': [true, 'always'], // 19.11
    'max-line-length': [true, 100], // 19.12
    'block-spacing': true, // 19.13
    'ter-computed-property-spacing': true, // 19.15
    'ter-func-call-spacing': true, // 19.16
    'no-trailing-whitespace': true, // 19.18
    'no-consecutive-blank-lines': true, // 19.19
    'trailing-comma': [
      true,
      {
        multiline: 'always',
        singleline: 'never',
        esSpecCompliant: true,
      },
    ], // 20.2
    semicolon: [true, 'always'], // 21.1
    'no-construct': true, // 22.2, 22.3, 22.6
    radix: true, // 22.3
    'function-name': [
      true,
      {
        'function-regex': /^[a-z$][\w\d]+$/,
        'method-regex': /^[a-z$][\w\d]+$/,
        'private-method-regex': /^[a-z$][\w\d]+$/,
        'protected-method-regex': /^[a-z$][\w\d]+$/,
        'static-method-regex': /^[a-z$][\w\d]+$/,
      }, // 23.1
    ],
    'variable-name': [true, 'check-format'], // 23.2
    'class-name': true, // 23.3
    'no-this-assignment': [true, 'allow-destructuring'], // 23.5
    'import-name': true, // 23.6
  },
};
